os/arch/arm/src/armv7-a/arm_mmu.c

```diff
@@ -254,3 +281,190 @@ void mmu_invalidate_region(uint32_t vstart, size_t size)
     }
 }
 #endif
+
+#ifdef CONFIG_APP_BINARY_SEPARATION
+/****************************************************************************
+ * Name: mmu_get_os_l1_pgtbl
+ *
+ * Description:
+ *   Returns the virtual address of the kernel L1 page table.
+ *
+ * Input Parameters:
+ *
+ * Returned Value:
+ * Page table address
+ ****************************************************************************/
+uint32_t *mmu_get_os_l1_pgtbl(void)
+{
+	return (uint32_t *)PGTABLE_BASE_VADDR;
+}
+
+/****************************************************************************
+ * Name: mmu_allocate_app_l1_pgtbl
+ *
+ * Description:
+ *   Allocate space for L1 page table of application, in accordance with
+ *   the requirements of the arch specific mmu.
+ *
+ * Input Parameters:
+ *
+ * Returned Value:
+ * L1 Page table address
+ ****************************************************************************/
+uint32_t *mmu_allocate_app_l1_pgtbl(int app_id)
+{
+	uint32_t *addr = (uint32_t *)(PGTABLE_BASE_VADDR + (app_id * 16384));
+	return addr;
+}
+
+/****************************************************************************
+ * Name: mmu_allocate_app_l2_pgtbl
+ *
+ * Description:
+ *   Allocate space for L2 page table of application, in accordance with
+ *   the requirements of the arch specific mmu.
+ *
+ * Input Parameters:
+ *
+ * Returned Value:
+ * L2 Page table address
+ ****************************************************************************/
+uint32_t *mmu_allocate_app_l2_pgtbl(int app_id, int l2_idx)
+{
+	uint32_t *addr = (uint32_t *)(PGTABLE_BASE_VADDR + (3 * 16384) + (app_id * 4096) + (l2_idx * 1024));
+	return addr;
+}
+
+/****************************************************************************
+ * Name: mmu_update_app_l1_pgtbl_ospgtbl
+ *
+ * Description:
+ * Loop through the L1 page table.
+ * Copy kernel L1 page table to app page table.
+ * If the entry is pointing to a L2 page table
+ * Allocate L2 page table for app.
+ * Copy entries from kernel to app L2 table.
+ * Update the L2 page table address in L1 table.
+ * 
+ * Input Parameters:
+ * app_pgtbl: Pointer to L1 page table of app
+ *
+ ****************************************************************************/
+void mmu_update_app_l1_pgtbl_ospgtbl(uint32_t *app_l1_pgtbl)
+{
+	uint32_t *os_l1_pgtbl = (uint32_t *)PGTABLE_BASE_VADDR;
+	memcpy((void *)app_l1_pgtbl, (void *)os_l1_pgtbl, L1_PGTBL_SIZE);
+}
+
+/****************************************************************************
+ * Name: mmu_map_app_region
+ *
+ * Description
+ *
+ * Input Parameters:
+ *
+ * Returned Value:
+ ****************************************************************************/
+void mmu_map_app_region(int app_id, uint32_t *l1_pgtbl, uint32_t start, uint32_t size, bool ro, bool exec, bool global)
+{
+	uint32_t idx;
+	uint32_t val;
+	uint32_t end = start + size;
+	int l2_idx = 0;
```

Comment:
```l2_idx``` is reset for every call to mmu_map_app_region and in this case, we call it thrice for rox, ro, and rw. So I think we may need to store it for an app.

Reply:
Fixed.